DOCS-1273: Operator 6.0.0 Deploy/Upgrade docs, removing Console references (#1284)
minio/docs: source/operations/install-deploy-manage/deploy-minio-tenant-helm.rst

The following procedure deploys a MinIO Tenant using the MinIO Operator Chart Repository.
This method supports a simplified installation path compared to the :ref:`local chart installation <deploy-tenant-helm-local>`.

The following procedure uses Helm to deploy a MinIO Tenant using the official MinIO Tenant Chart.

.. important::

   If you use Helm to deploy a MinIO Tenant, you must use Helm to manage or upgrade that deployment.

>    Do not use ``kubectl krew``, Kustomize, or similar methods to manage or upgrade the MinIO Tenant.

Author: I think keep it because it _was_ around at one point. We want to emphasize that even if someone was using or heard of or read of using `krew`, its no longer recommended.

   .. code-block:: shell
      :class: copyable
      :substitutions:

      NAME                            CHART VERSION   APP VERSION     DESCRIPTION                    
      minio-operator/minio-operator   4.3.7           v4.3.7          A Helm chart for MinIO Operator
      minio-operator/operator         |operator-version-stable|           v|operator-version-stable|          A Helm chart for MinIO Operator
      minio-operator/tenant           |operator-version-stable|           v|operator-version-stable|          A Helm chart for MinIO Operator

#. Create a local copy of the Helm ``values.yaml``

   Download the base ``values.yaml`` object for modification:

   .. code-block:: shell
      :class: copyable

      curl -sLo values.yaml https://raw.githubusercontent.com/minio/operator/master/helm/tenant/values.yaml

>    Open the ``values.yaml`` object in your preferred text edit.

Reviewer: ```suggestion
   Open the ``values.yaml`` object in your preferred text editor.
```

          If no storage class exists matching the specified value *or* if the specified storage class cannot meet the requested number of PVCs or storage capacity, the Tenant may fail to start.

      * - ``size``
        - The amount of storage to request for each generated PVC.

#. Configure Tenant Affinity or Anti-Affinity

   The Tenant Chart supports the following Kubernetes Selector, Affinity and Anti-Affinity configurations:

   - Node Selector (``tenant.nodeSelector``)
   - Node/Pod Affinity or Anti-Affinity (``spec.pools[n].affinity``)

   MinIO recommends configuring Tenants with Pod Anti-Affinity to ensure that the Kubernetes schedule does not schedule multiple pods on the same worker node.

   If you have specific worker nodes on which you want to deploy the tenant, pass those node labels or filters to the ``nodeSelector`` or ``affinity`` field to constrain the scheduler to place pods on those nodes.

#. Configure Network Encryption

   The MinIO Tenant CRD provides the following fields from which you can configure tenant TLS network encryption:

   .. list-table::
      :header-rows: 1
      :widths: 30 70

      * - Field
        - Description

>       * - ``tenant.certificate.requestAutoCert``
>         - Enables or disables MinIO :ref:`automatic TLS certificate generation <minio-tls>`

Reviewer: are the accepted values here `true` or `false`?
```suggestion
      * - ``tenant.certificate.requestAutoCert``
        - Enable or disable MinIO :ref:`automatic TLS certificate generation <minio-tls>`
```

Reviewer 2: Yes @feorlen

Reviewer 2: @ravindk89 @feorlen Are we using the imperative mood or simply describing the field (indicative mood)? It's not clear what the standard is.

Author: My intent was to be imperative (I think), but not necessarily because it's our standard - it's just how I learned to write these sort of things.

In which case i think we'd leave this as-is. Maybe for docs to discuss.

Author: I changed everything and lets see how we feel about it.

      * - ``tenant.certificate.certConfig``
        - Controls the settings for :ref:`automatic TLS <minio-tls>`.

>           Requires ``spec.requestAutoCert: true``

Reviewer: "Controls" is a bit vague, is it worth mentioning the type of accepted values here? There is more info in the CRD, which is reasonable to presume the reader is looking at.

Reviewer 2: Also `spec.requestAutoCert` can be `true`, `false`, or simply not present. It is not a required field.

Author: @allanrogerr iirc if `requestAutoCert`` is ``false`` then changes to this field do nothing, right?

Indeed omitting `requestAutoCert` defaults to `true` behavior so I could reverse this

```suggestion
      * - ``tenant.certificate.certConfig``
        - Controls the settings for :ref:`automatic TLS <minio-tls>`.
          Only has affect if ``spec.requestAutoCert`` is not false.
```

Author: The easiest path here is to just state 'controls he settings...if enabled` and let the user infer that disabling one thing disables the othe.r

      * - ``tenant.certificate.externalCertSecret``
        - Specify one or more Kubernetes secrets of type ``kubernetes.io/tls`` or ``cert-manager``.
          MinIO uses these certificates for performing TLS handshakes based on hostname (Server Name Indication).

      * - ``tenant.certificate.externalCACertSecret``
        - Specify one or more Kubernetes secrets of type ``kubernetes.io/tls`` with the Certificate Authority (CA) chains which the Tenant must trust for allowing client TLS connections.

#. Configure MinIO Environment Variables

   You can set MinIO Server environment variables using the ``tenant.configuration`` field.

   The field must specify a Kubernetes opaque secret whose data payload ``config.env`` contains each MinIO environment variable you want to set.

   The YAML includes an object ``kind: Secret`` with ``metadata.name: storage-configuration`` that sets the root username, password, erasure parity settings, and enables Tenant Console.

>    Modify this as-needed to reflect your Tenant requirements.

Reviewer: ```suggestion
   Modify this as needed to reflect your Tenant requirements.
```

#. Deploy the Tenant

   Use ``helm`` to install the Tenant Chart using your ``values.yaml`` as an override:

   .. code-block:: shell
      :class: copyable

      helm install \

>         --namespace minio-tenant-1 \

Reviewer 2: Not sure if you want to use `MINIO_TENANT_NAMESPACE` and `MINIO_TENANT_NAME` nomenclature, but we should be consistent.

Author: Hm. Yeah we kind of go back and forth on this. I'll migrate to the var-replace style.

        --create-namespace \
        --values values.yaml \

>         tenant minio/tenant

Reviewer 2: ```suggestion
        tenant minio-operator/tenant
```

   You can monitor the progress using the following command:

   .. code-block:: shell
      :class: copyable

>       watch kubectl get all -n minio-tenant

Reviewer 2: ```suggestion
      watch kubectl get all -n minio-tenant-1
```

Reviewer 2: Not sure if you want to use `MINIO_TENANT_NAMESPACE` and `MINIO_TENANT_NAME` nomenclature, but we should be consistent.

   Each chart contains a ``values.yaml`` file you can customize to suit your needs.
   For details on the options available in the MinIO Tenant ``values.yaml``, see :ref:`minio-tenant-chart-values`.
   
   Open the ``values.yaml`` object in your preferred text editor.

#. Configure the Tenant topology
   
   The following fields share the ``tenant.pools[0]`` prefix and control the number of servers, volumes per server, and storage class of all pods deployed in the Tenant:
   
   .. list-table::
      :header-rows: 1
      :widths: 30 70

      * - Field
        - Description

      * - ``servers`` 
        - The number of MinIO pods to deploy in the Server Pool.
      * - ``volumesPerServer`` 
        - The number of persistent volumes to attach to each MinIO pod (``servers``).
          The Operator generates ``volumesPerServer x servers`` Persistant Volume Claims for the Tenant.
      * - ``storageClassName`` 
        - The Kubernetes storage class to associate with the generated Persistent Volume Claims.

>           If no storage class exists matching the specified value *or* if the specified storage class cannot meet the requested number of PVCs or storage capacity, the Tenant may fail to start.

Author: We specify in the Persistent Volumes pre-reqs that we expect the PV to exist already. I'll make that more explicit.

   .. list-table::
      :header-rows: 1
      :widths: 30 70

      * - Field
        - Description

      * - ``tenant.certificate.requestAutoCert``
        - Enables or disables MinIO :ref:`automatic TLS certificate generation <minio-tls>`

      * - ``tenant.certificate.certConfig``
        - Controls the settings for :ref:`automatic TLS <minio-tls>`.
          Requires ``spec.requestAutoCert: true``

      * - ``tenant.certificate.externalCertSecret``
        - Specify one or more Kubernetes secrets of type ``kubernetes.io/tls`` or ``cert-manager``.
          MinIO uses these certificates for performing TLS handshakes based on hostname (Server Name Indication).

      * - ``tenant.certificate.externalCACertSecret``
        - Specify one or more Kubernetes secrets of type ``kubernetes.io/tls`` with the Certificate Authority (CA) chains which the Tenant must trust for allowing client TLS connections.

#. Configure MinIO Environment Variables

   You can set MinIO Server environment variables using the ``tenant.configuration`` field.

   The field must specify a Kubernetes opaque secret whose data payload ``config.env`` contains each MinIO environment variable you want to set.

>    The YAML includes an object ``kind: Secret`` with ``metadata.name: storage-configuration`` that sets the root username, password, erasure parity settings, and enables Tenant Console.

Author: Made a pass.

   Each chart contains a ``values.yaml`` file you can customize to suit your needs.

>    For details on the options available in the MinIO Tenant ``values.yaml``, see :ref:`minio-tenant-chart-values`.

Reviewer: Is this link still needed if we are now actually discussing the contents of values.yaml here?

Author: I think so - `values.yaml` has a lot more gritty detail, this tutorial (for now) is very high level.

>    Open the ``values.yaml`` object in your preferred text edit.

Reviewer: ```suggestion
   Open the ``values.yaml`` object in your preferred text editor.
```
